### DISCUSSION TITLE
IDL generated code suggestions [13724]

### DISCUSSION THREAD

**Discussion by jwillemsen** [eProsima/Fast-DDS-Gen#240] [Ideas]
When looking at the IDL generated code (for example https://github.com/eProsima/Fast-DDS/blob/master/examples/C%2B%2B/Keys/sample.h) I have the following suggestions:
- Make use of C++11 uniform initialization (see for example) https://www.geeksforgeeks.org/uniform-initialization-in-c/ to initialize all members, that way the constructor can just be default
- Make use of the default feature for copy/move assignment/constructor, only unions need some special handling, for structs there is nothing special so the compiler default should work
- isKeyDefined() could maybe be a std::true_type or std::false_type, see https://stackoverflow.com/questions/58694521/what-is-stdfalse-type-or-stdtrue-type and also the IDL to C++11 language mapping uses that for its IDL traits
- destructor can be default in most cases
- there are some static methods generated which are fastdds specific, maybe move them to a separate templated struct so they are clearly separated, the user doesn't want to see those when he looks at his idl defined type
- this example doesn't use any array/vector, so why add includes for that to the header?
- the class has no explicit constructor accepting values for each struct member in the order they are specified in IDL
- there is no namespace level swap generated as defined by IDL to C++11

**Comment by jwillemsen** [comment]
See https://github.com/RemedyIT/idl2cpp11/blob/master/examples/struct/testC.h for some ideas, that file was generated using TAOX11 (http://www.taox11.org)

**Comment by jwillemsen** [comment]
#39 is about unions, these can benefit from several of the suggestions in this issue also

**Comment by JLBuenoLopez** [comment]
Thanks for the suggestions @jwillemsen 

We are open to external contributions so feel free to open the corresponding PRs if you want to. I am also moving this ticket to the Ideas discussion forum where enhancements and feature requests are discussed.
